rfd/0abc-kube-access-forward-identity.md

```diff
@@ -0,0 +1,205 @@
+	// so that new logins with different parameters will not reuse this context
+	return fmt.Sprintf("%v:%v:%v:%v:%v:%v:%v", c.teleportCluster.name, c.User.GetName(), c.kubeUsers, c.kubeGroups, c.kubeClusterName, c.certExpires.Unix(), c.Identity.GetIdentity().ActiveRequests)
+}
+```
+
+Each time we introduce a new field to the user's identity, we need to make sure
+that the key is updated to include the new field if the field is used by Kubernetes
+Access. This process is error-prone and it's easy to forget to update the key, and
+eventually introduce a bug or a security vulnerability.
+
+This RFD proposes a way to forward the user's identity from the proxy to the
+Kube Service and remote Teleport proxy that does not require the proxy to generate
+a new certificate for each user's identity.
+
+## Details
+
+The Proxy will authenticate to the Kube Service and any remote Teleport proxy using its
+certificate. The certificate used by proxy does not contain any user identification
+and it's signed by the Teleport CA. The proxy will then forward the user's identity
+using HTTP headers or gRPC metadata instead of generating a new certificate on
+the behalf of the user. Since the proxy is authenticated using its certificate,
+the upstream service can trust the identity received in the request headers.
+
+This change will remove the requirement for the proxy to generate a new certificate
+and thus call auth's `ProcessKubeCSR` endpoint to sign the key-pair containing the
+user's identity.
+
+
+### Impersonation: Forward user identity using HTTP headers
+
+As opposed to the majority of protocols supported by Teleport, the Kubernetes protocol
+is based on HTTP 1/2. This allows us to forward the user's identity using HTTP headers
+for the specific case of Kubernetes.
+
+The proposed solution is to use the Proxy certificate to authenticate to the upstream
+service and to forward the user's identity using HTTP headers or gRPC metadata
+for later Impersonation.
+
+Once the upstream service receives a request, it will check the certificate provided
+to validate the request's provenance and its role - making sure it's a proxy - for
+authenticating the request. Once authenticated, it will impersonate the identity
+contained in the request headers. This impersonated identity will be available in
+`authz.Context` and services won't be able to distinguish between a request forwarded
+by the proxy or a request made directly by the user.
+
+This approach is similar to what Kubernetes API does when
+forwarding requests to the API server using Impersonation. In the Teleport case, we must
+forward the full user's identity instead of just the username/roles.
+
+Since the proxy has full access to the HTTP request, it can add the user's identity
+to the request headers before forwarding it to the upstream service.
+
+```go
+headers["TELEPORT_IMPERSONATE_IDENTITY"] = clientCert.Subject
```

Comment:
What format will the header value be? base64 encoded or just plain json-encoded identity object?